### PR TITLE
Implement rate limiting

### PR DESCRIPTION
**Pull request by SibrenVasse**
Based on #199 #203. 
Implementation for rate limiting, specified limit is per IP.
- Puts limit on endpoints like username/password authentication and mfa endpoints.
- Only counts 'bad requests', e.g. failed login attempt.
- I arbitrarily chose 10/min limit. Any thoughts? As it's only failed requests, we could possibly lower this. More complex limits are possible:
```
@limiter.limit("100/day;10/hour;1/minute")
```

Nginx snippet
```
server {
    server_name  app.mydomain.com;

    location / {
        proxy_pass http://localhost:7777;
        proxy_set_header X-Forwarded-For $remote_addr;
    }
}
```

### CODE REVIEW

**Comment by developStorm** [comment]
It's ok for giving webauthn more quotas since it's nearly impossible to be brute-force attacked

**Comment by SibrenVasse** [comment]
Yeah but is there really a good reason to fail the webauthn check 10 times in a minute? How would you manage that.

**Comment by developStorm** [comment]
Oops I were thought the "more complex example" is the actual rule (1/min)😅10/min sounds good!

File: app/api/views/auth.py [hunk 1]
```diff
@@ -164,6 +178,9 @@ def auth_activate():
 
 @api_bp.route("/auth/reactivate", methods=["POST"])
 @cross_origin()
+@limiter.limit(
```
Comment (nguyenkims) [inline app/api/views/auth.py:181]: the `limiter` is currently not used in this endpoint right?

File: app/auth/views/forgot_password.py [hunk 2]
```diff
@@ -12,6 +13,9 @@ class ForgotPasswordForm(FlaskForm):
 
 
 @auth_bp.route("/forgot_password", methods=["GET", "POST"])
+@limiter.limit(
+    "10/minute", deduct_when=lambda r: hasattr(g, "forgot_fail") and g.forgot_fail
```
Comment (nguyenkims) [inline app/auth/views/forgot_password.py:17]: `forgot_fail` should be renamed to `deduct_limit` here right :) ?

File: templates/error/429.html [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{% extends "error.html" %}
+
+{% block error_name %}
+  429
+{% endblock %}
+
+{% block error_description %}
+  Slow down there, pardner!
```
Comment (nguyenkims) [inline templates/error/429.html:8]: 😅

File: server.py [hunk 4]
```diff
@@ -369,6 +371,13 @@ def forbidden(e):
         else:
             return render_template("error/403.html"), 403
 
+    @app.errorhandler(429)
+    def forbidden(e):
+        if request.path.startswith("/api/"):
```
Comment (nguyenkims) [inline server.py:376]: can we add a `log.error` here please? We got paged when an error happens so this allows us to see either an attack is going on or is this rather a false positive.

File: app/limiter.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+from flask import request
+from flask_limiter import Limiter
+from flask_limiter.util import get_remote_address
+
+# Setup rate limit facility
+limiter = Limiter(key_func=get_remote_address)
```
[the remark this answers is not in the document]
Comment (SibrenVasse) [reply app/limiter.py]: Ah yes, that's the location I was looking for but failed to find ;)

**Pull request merged**